File: web/src/lib/dayDetail.ts
import { niceStep } from "./normals";
import type { DailyForecast, HourlyForecast } from "./weatherApi";

// Buckets the hourly window by local calendar day (YYYY-MM-DD in timeZone).
export function groupHoursByLocalDate(
  hours: readonly HourlyForecast[],
  timeZone: string,
): Map<string, HourlyForecast[]> {
  const formatter = new Intl.DateTimeFormat("en-CA", {
    timeZone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  });
  const groups = new Map<string, HourlyForecast[]>();
  for (const hour of hours) {
    const date = formatter.format(new Date(hour.time));
    const bucket = groups.get(date);
    if (bucket) {
      bucket.push(hour);
    } else {
      groups.set(date, [hour]);
    }
  }
  return groups;
}

export type DaySummary = {
  high: number | null;
  low: number | null;
  precipitation: number | null;
  popMax: number | null;
  windAvg: number | null;
  gustMax: number | null;
  humidityAvg: number | null;
  cloudCoverAvg: number | null;
};

// High and low come from the daily row the user clicked so the sheet repeats
// the numbers they saw. Precipitation is the total over the hours shown, so
// it matches the histogram even for today, whose daily row only covers the
// hours after the forecast was fetched. The rest is derived from the hours.
export function summarizeDay(
  day: DailyForecast,
  hours: readonly HourlyForecast[],
): DaySummary {
  const temps = defined(hours.map((h) => h.temperature));
  return {
    high: day.high ?? maxOf(temps),
    low: day.low ?? minOf(temps),
    precipitation:
      sumOf(defined(hours.map((h) => h.precipitation_1h))) ??
      day.precipitation_mm ??
      day.precipitation_1h_sum ??
      null,
    popMax: maxOf(defined(hours.map((h) => h.pop))),
    windAvg:
      day.wind_speed_avg ?? avgOf(defined(hours.map((h) => h.wind_speed))),
    gustMax:
      day.hourly_maximum_gust_max ??
      maxOf(defined(hours.map((h) => h.wind_gust))),
    humidityAvg:
      day.humidity_avg ?? avgOf(defined(hours.map((h) => h.humidity))),
    cloudCoverAvg:
      day.total_cloud_cover_avg ??
      avgOf(defined(hours.map((h) => h.cloud_cover))),
  };
}

export type AxisTick = { value: number; y: number };

export type TemperatureChart = {
  min: number;
  max: number;
  linePath: string;
  feelsPath: string;
  ticks: AxisTick[];
};

export type PrecipitationChart = {
  max: number;
  bars: { x: number; width: number; height: number; value: number }[];
  ticks: AxisTick[];
};

export type DayChart = {
  temperature: TemperatureChart | null;
  precipitation: PrecipitationChart;
  hourLabels: { x: number; label: string; major: boolean }[];
};

// The hour axis is always the full local day, so a chart for today shows
// the elapsed hours as empty space and every day lines up the same way.
export const HOUR_LABELS: DayChart["hourLabels"] = Array.from(
  { length: 8 },
  (_, i) => {
    const hour = i * 3;
    return {
      x: slotX(hour),
      label: String(hour).padStart(2, "0"),
      major: hour % 6 === 0,
    };
  },
);

function slotX(hour: number): number {
  return ((hour + 0.5) / 24) * 100;
}

// Temperature/feels-like curves and a precipitation histogram for one day,
// each in percentages of its own plot box and sharing the hour axis.
// `localHour` gives an entry's hour of day (0–23) in the place's timezone.
export function buildDayChart(
  hours: readonly HourlyForecast[],
  localHour: (time: string) => number,
): DayChart | null {
  if (hours.length < 2) return null;
  const x = (index: number) => slotX(localHour(hours[index].time));
  return {
    temperature: temperatureChart(hours, x),
    precipitation: precipitationChart(hours, x),
    hourLabels: HOUR_LABELS,
  };
}

function temperatureChart(
  hours: readonly HourlyForecast[],
  x: (index: number) => number,
): TemperatureChart | null {
  const temps = defined(hours.map((h) => h.temperature));
  if (temps.length < 2) return null;
  const values = [...temps, ...defined(hours.map((h) => h.feels_like))];
  const min = Math.floor(Math.min(...values)) - 1;
  const max = Math.ceil(Math.max(...values)) + 1;
  const y = (value: number) => ((value - min) / (max - min)) * 100;

  const path = (pick: (hour: HourlyForecast) => number | null | undefined) => {
    const parts: string[] = [];
    hours.forEach((hour, i) => {
      const value = pick(hour);
      if (value == null) return;
      parts.push(
        `${parts.length === 0 ? "M" : "L"}${fmt(x(i))},${fmt(100 - y(value))}`,
      );
    });
    return parts.length > 1 ? parts.join("") : "";
  };

  return {
    min,
    max,
    linePath: path((h) => h.temperature),
    feelsPath: path((h) => h.feels_like),
    ticks: axisTicks(min, max, 4, y),
  };
}

// The histogram always renders, so a dry day reads as an empty 0–1 mm plot
// rather than a missing one.
function precipitationChart(
  hours: readonly HourlyForecast[],
  x: (index: number) => number,
): PrecipitationChart {
  const precip = hours.map((h) => h.precipitation_1h ?? 0);
  const step = niceStep(Math.max(...precip, 1) / 2);
  const max = Math.max(1, Math.ceil(Math.max(...precip, 0) / step) * step);
  const y = (value: number) => (value / max) * 100;
  const width = (100 / 24) * 0.7;
  const bars: PrecipitationChart["bars"] = [];
  precip.forEach((value, i) => {
    if (value <= 0) return;
    bars.push({ x: x(i), width, height: y(value), value });
  });
  return { max, bars, ticks: axisTicks(step, max, 0, y, step) };
}

function axisTicks(
  min: number,
  max: number,
  divisions: number,
  y: (value: number) => number,
  step = niceStep((max - min) / divisions),
): AxisTick[] {
  const ticks: AxisTick[] = [];
  for (let value = Math.ceil(min / step) * step; value <= max; value += step) {
    ticks.push({ value, y: y(value) });
  }
  return ticks;
}

function defined(values: readonly (number | null | undefined)[]): number[] {
  return values.filter((value): value is number => value != null);
}

function maxOf(values: number[]): number | null {
  return values.length > 0 ? Math.max(...values) : null;
}

function minOf(values: number[]): number | null {
  return values.length > 0 ? Math.min(...values) : null;
}

function sumOf(values: number[]): number | null {
  return values.length > 0 ? values.reduce((a, b) => a + b, 0) : null;
}

function avgOf(values: number[]): number | null {
  const sum = sumOf(values);
  return sum == null ? null : sum / values.length;
}

function fmt(value: number): string {
  return String(Math.round(value * 100) / 100);
}

File: ios/wby/wby/Models/DayDetail.swift
import Foundation

/// One forecast day together with its hours in the place's local calendar
/// day, plus the summary numbers the day sheet shows.
struct DayDetail: Identifiable {
    let day: DailyForecast
    let hours: [HourlyForecast]
    let timeZone: TimeZone

    var id: String { day.date }

    /// Pairs every daily forecast with the hours falling on its local date.
    static func build(days: [DailyForecast], hours: [HourlyForecast], timeZone: TimeZone) -> [DayDetail] {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = timeZone
        var byDate: [String: [HourlyForecast]] = [:]
        for hour in hours.sorted(by: { $0.time < $1.time }) {
            byDate[formatter.string(from: hour.time), default: []].append(hour)
        }
        return days.map { DayDetail(day: $0, hours: byDate[$0.date] ?? [], timeZone: timeZone) }
    }

    // MARK: - Dates

    private var calendar: Calendar {
        var calendar = Calendar.current
        calendar.timeZone = timeZone
        return calendar
    }

    var date: Date? { day.displayDate(timeZone: timeZone) }

    var isToday: Bool {
        guard let date else { return false }
        return calendar.isDateInToday(date)
    }

    /// Short name for the toolbar: "Today", "Tomorrow" or the weekday.
    var shortTitle: String {
        guard let date else { return day.date }
        if calendar.isDateInToday(date) { return "Today" }
        if calendar.isDateInTomorrow(date) { return "Tomorrow" }
        return Self.weekdayFormatter(timeZone).string(from: date)
    }

    /// Full date for the sheet header, e.g. "Wednesday 9 September".
    var longTitle: String {
        guard let date else { return day.date }
        return Self.longDateFormatter(timeZone).string(from: date)
    }

    var conditionDescription: String {
        WeatherSymbols.conditionDescription(from: day.symbol) ?? "Forecast"
    }

    /// Hour of day (0–23) of an hourly entry in the place's timezone.
    func localHour(of date: Date) -> Int {
        calendar.component(.hour, from: date)
    }

    /// Where "now" sits in the day as a 0–1 fraction; nil unless this is today.
    var nowFraction: Double? {
        guard isToday else { return nil }
        let now = Date()
        let hour = Double(calendar.component(.hour, from: now))
        let minute = Double(calendar.component(.minute, from: now))
        return (hour + minute / 60) / 24
    }

    // MARK: - Summary

    /// High and low come from the daily row so the sheet repeats the numbers
    /// the user tapped. Precipitation is the total over the hours shown so it
    /// matches the histogram even for today. The rest is derived from hours
    /// where the daily row has nothing.
    var high: Double? { day.high ?? hours.compactMap(\.temperature).max() }
    var low: Double? { day.low ?? hours.compactMap(\.temperature).min() }
    var precipitation: Double? {
        let hourly = hours.compactMap(\.precipitation1h)
        if !hourly.isEmpty { return hourly.reduce(0, +) }
        return day.precipitationMm ?? day.precipitation1hSum
    }
    var popMax: Double? { hours.compactMap(\.pop).max() }
    var windAvg: Double? { day.windSpeedAvg ?? Self.average(hours.compactMap(\.windSpeed)) }
    var gustMax: Double? { day.hourlyMaximumGustMax ?? hours.compactMap(\.windGust).max() }
    var humidityAvg: Double? { day.humidityAvg ?? Self.average(hours.compactMap(\.humidity)) }
    var cloudCoverAvg: Double? { day.totalCloudCoverAvg ?? Self.average(hours.compactMap(\.cloudCover)) }

    struct Stat: Identifiable {
        let label: String
        let value: String

        var id: String { label }
    }

    var stats: [Stat] {
        [
            Stat(label: "High", value: Self.formatTemp(high)),
            Stat(label: "Low", value: Self.formatTemp(low)),
            Stat(label: "Precipitation", value: Self.formatPrecip(precipitation)),
            Stat(label: "Chance of rain", value: Self.formatPercent(popMax)),
            Stat(label: "Wind", value: Self.formatSpeed(windAvg)),
            Stat(label: "Gusts", value: Self.formatSpeed(gustMax)),
            Stat(label: "Humidity", value: Self.formatPercent(humidityAvg)),
            Stat(label: "Cloud cover", value: Self.formatPercent(cloudCoverAvg)),
        ]
    }

    private static func average(_ values: [Double]) -> Double? {
        guard !values.isEmpty else { return nil }
        return values.reduce(0, +) / Double(values.count)
    }

    // MARK: - Formatting

    static func formatTemp(_ value: Double?) -> String {
        guard let value else { return "--" }
        return "\(Int(value.rounded()))°"
    }

    static func formatPrecip(_ value: Double?) -> String {
        guard let value else { return "--" }
        if abs(value.rounded() - value) < 0.05 {
            return "\(Int(value.rounded())) mm"
        }
        return String(format: "%.1f mm", value)
    }

    static func formatPercent(_ value: Double?) -> String {
        guard let value else { return "--" }
        return "\(Int(value.rounded()))%"
    }

    static func formatSpeed(_ value: Double?) -> String {
        guard let value else { return "--" }
        return "\(Int(value.rounded())) m/s"
    }

    private static func weekdayFormatter(_ timeZone: TimeZone) -> DateFormatter {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE"
        formatter.timeZone = timeZone
        return formatter
    }

    private static func longDateFormatter(_ timeZone: TimeZone) -> DateFormatter {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE d MMMM"
        formatter.timeZone = timeZone
        return formatter
    }
}
